control-plane/controller/api-gateway/consul/config_entry_translation.go

```diff
@@ -0,0 +1,100 @@
+package consul
+
+import (
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/hashicorp/consul/api"
+	consulAPI "github.com/hashicorp/consul/api"
+)
+
+const (
+	metaKeyManagedBy       = "managed-by"
+	metaValueManagedBy     = "consul-k8s-gateway-controller"
+	metaKeyKubeNS          = "k8s-namespace"
+	metaKeyKubeServiceName = "k8s-service-name"
+)
+
+func GatewayToAPIGateway(k8sGW gwv1beta1.Gateway) consulAPI.APIGatewayConfigEntry {
+	listeners := make([]consulAPI.APIGatewayListener, 0, len(k8sGW.Spec.Listeners))
+	conditions := make([]consulAPI.Condition, 0, len(k8sGW.Status.Conditions))
+	for _, listener := range k8sGW.Spec.Listeners {
+		certificates := make([]consulAPI.ResourceReference, 0, len(listener.TLS.CertificateRefs))
+		for _, certificate := range listener.TLS.CertificateRefs {
+			c := consulAPI.ResourceReference{
+				Kind:        consulAPI.InlineCertificate,
+				Name:        string(certificate.Name),
+				SectionName: "",
+				Partition:   "",
+				Namespace:   string(*certificate.Namespace),
+			}
+			certificates = append(certificates, c)
+		}
+		l := consulAPI.APIGatewayListener{
+			Name:     string(listener.Name),
+			Hostname: string(*listener.Hostname),
+			Port:     int(listener.Port),
+			Protocol: string(listener.Protocol),
+			TLS: consulAPI.APIGatewayTLSConfiguration{
+				Certificates: certificates,
+			},
+		}
+
+		listeners = append(listeners, l)
+	}
+
+	for _, condition := range k8sGW.Status.Conditions {
+		// TODO: John Maguire - convert status/reason/type to use a mapping of values defined
+		// in consul to convert from the k8s status/type/reason into consul status/types/reason
+		c := consulAPI.Condition{
+			Type:    condition.Type,
+			Status:  string(condition.Status),
+			Reason:  condition.Reason,
+			Message: condition.Message,
+			Resource: &consulAPI.ResourceReference{
+				Kind:      consulAPI.APIGateway,
+				Name:      k8sGW.Name,
+				Namespace: k8sGW.Namespace,
+			},
+			LastTransitionTime: &condition.LastTransitionTime.Time,
+		}
+		conditions = append(conditions, c)
+	}
+
+	for _, listener := range k8sGW.Status.Listeners {
+		for _, condition := range listener.Conditions {
+			listenerCondition := consulAPI.Condition{
+				Type:    condition.Type,
+				Status:  string(condition.Status),
+				Reason:  condition.Reason,
+				Message: condition.Message,
+				Resource: &consulAPI.ResourceReference{
+					Kind:        consulAPI.APIGateway,
+					Name:        k8sGW.Name,
+					SectionName: string(listener.Name),
+					Namespace:   k8sGW.Namespace,
+				},
+				LastTransitionTime: &condition.LastTransitionTime.Time,
+			}
+			conditions = append(conditions, listenerCondition)
+		}
+	}
+
+	return consulAPI.APIGatewayConfigEntry{
```

Comment:
how do we support admin partitions here? is there something in the k8s definition that would hint to use what admin partition this should be in? also would we be worried about if a user was not mirroring consul namespaces/do we have a way to check that at runtime?

Reply:
So, the admin partitions are something that a consul-k8s installation handles globally. Anything we synchronize over to Consul will wind up in a single partition since we set the Consul client in our deployment to use a single partition via env variable:

https://github.com/hashicorp/consul-k8s/blob/11f842cbe94b5df544c03e2441a1820397f92b9b/charts/consul/templates/_helpers.tpl#L309-L311

If you want to be explicit about writing the partition value in, you'll want to pull that out of the env variable, otherwise you might be able to rely on the defaulting normalization behavior in our Consul entries (though that may make some of the dirty-checking a bit more complex for determining whether we need to actually write back to Consul or not).